core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java

```diff
@@ -250,33 +277,71 @@ public void initializeDefaultCodec() {
        */
       @Override
       protected CompressionCodec createNewCodec(final int bufferSize) {
-        DefaultCodec myCodec = new DefaultCodec();
-        Configuration myConf = new Configuration(conf);
-        // only use the buffersize if > 0, otherwise we'll use
-        // the default defined within the codec
-        if (bufferSize > 0)
-          myConf.setInt(BUFFER_SIZE_OPT, bufferSize);
-        myCodec.setConf(myConf);
-        return myCodec;
+        if (USE_QAT) {
+          String extClazz =
+              (conf.get(CONF_QAT_CLASS) == null ? System.getProperty(CONF_QAT_CLASS) : null);
+          String clazz = (extClazz != null) ? extClazz : DEFAULT_QAT_CLASS;
+          try {
+            LOG.info("Trying to load qat codec class: " + clazz);
+
+            Configuration myConf = new Configuration(conf);
+            // only use the buffersize if > 0, otherwise we'll use
+            // the default defined within the codec
+            if (bufferSize > 0)
+              myConf.setInt(QAT_BUFFER_SIZE_OPT, bufferSize);
+
+            CompressionCodec c =
+                (CompressionCodec) ReflectionUtils.newInstance(Class.forName(clazz), myConf);
+            if (c instanceof Configurable) {
+              ((Configurable) c).setConf(myConf);
+            }
+            return c;
+          } catch (ClassNotFoundException e) {
+            LOG.error("Unable to create QAT codec", e);
+          }
+          return null;
```

Comment:
If this is null, should we fall back to the default impl? This seems like it would result in NPEs and break things in weird ways. The QuickAssist impl should probably be an optional optimization that works when available, but is ignored when not available.

Reply:
I was thinking that if someone specified `-Duse.qat`, then they were intending on using the device and expected it to work. The fact that it was not working in this case would be an error as maybe they have a configuration issue. Falling back to the default implementation and logging it is not honoring the users intent.

Reply:
Seems reasonable. Perhaps we can throw an exception instead of `return null`, so this error is explicit? Would want to make sure it is thrown early.

Reply:
On the other hand, we should weigh the user's intent to use GZ, whether or not the optimization is available. The user has two intents, to use GZ, and to use QAT. The current behavior says if QAT isn't available, then GZ isn't available at all. I'm not sure that's a good experience for users either.

Reply:
Yeah, I understand the desire to fail fast instead of diagnose when things aren't working as expected. I'm just thinking we have to infer what the user is expecting. I'm sure some users might expect "I want QAT's impl of GZ only". But other users might expect "I want GZ, QAT to make it faster, if available". Personally, I'd be somebody in the latter camp.

Could support both sets of user expectations with values of `.enabled=[true,false,required/force]`.